### PR TITLE
[design] Cluster ux long term vision document

### PR DESCRIPTION
**Pull request by chaas**
Design doc for cluster UX long term vision.

### CODE REVIEW

**Comment by benesch** [comment]
Cross-referencing the epic: #22120.

File: doc/developer/design/20231002_cluster_vision.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+We should also be more prescriptive about what data should be colocated, \
+e.g. when should the user create a new cluster for their new sources/MVs/indexes versus increase the size of their existing cluster.
+
+We believe this will make it clearer how to achieve appropriate fault tolerance and maxmimize resource efficiency.
+
+### Support & testing
+Support is able to create create unbilled or partially billed cluster resources for resolving customer issues. This is soon to be possible via unbilled replicas [#20317](https://github.com/MaterializeInc/materialize/issues/20317).
+
+Engineering is also able to create additional unbilled shadow replicas for testing new features and query plan changes, which do not serve customers' production workflows.
```
Comment (antiguru) [inline doc/developer/design/20231002_cluster_vision.md:79]: I'm not yet fully convinced if shadow replicas are the mechanism we'd like to have to do A/B testing. An alternative are shadow environments where all parts are cloned and we don't risk taking down the environment through a misbehaving shadow replica. All I mean to say is that we might want to leave it outside of this design!

Comment (chaas) [reply doc/developer/design/20231002_cluster_vision.md:79]: Sounds good, I can leave this out then if we're not sure yet

Comment (chaas) [reply doc/developer/design/20231002_cluster_vision.md:79]: I can leave out the entire "Support & testing" section if the content there is too narrow of a view of how we can support customers in the long term

**Comment by benesch** [comment]
In my queue to review! Likely not until Friday or the weekend though.

**Comment by sthm** [comment]
Moving towards the declarative API for managing clusters seems like the right thing to do. It's much easier to understand for customers and simpler to handle. But I would be a bit careful to deprecate the imperative API for customers. It seems reasonable to (almost completely) hide it in the documentation but I think there legit are use cases where customers can still benefit from it.

Here are two examples that require replicas with different sizes: For instance, it's an easy way to verify the effect of scaling up/down clusters. Right now, a customer I'm working with is seeing hydration times around 30min but the replica can potentially be scaled down. Adding an additional smaller replica is a very seamless way to just test what happens. Similarly, cost cautious customers may be willing to run replicas pretty busy, if they want to verify upfront if a smaller replica does not fail over they can provision the smaller replica in addition to the larger one and monitor for some time (hours or few days) what happens.

People may be able to do similar things with automatic and 0 down time scaling. I'm not saying that we should not focus on the declarative approach. But keeping some basic support for the more manual imperative approach could be useful for some specific and less common use cases.

File: doc/developer/design/20231002_cluster_vision.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+The primary work item for this is **graceful reconfiguration**. At the moment, a change in size causes downtime until the new replicas are hydrated. As such, customers still want the flexibility to create their own replicas for graceful resizing. We can avoid this by leaving a subset of the original replicas around until the new replicas are hydrated. \
+This requires us to 1) detect when hydration is complete and 2) trigger database object changes based on this event (without/based on an earlier DDL statement).
+
+Another consideration is internal use-cases, such as unbilled replicas. We may want to keep around an imperitive API for internal (support) use only.
+
+To be determined: whether replica sets fits into this model, either externally exposed or internal-only. Perhaps they are a way we could recover clusters with heterogeneous replicas while retaining a declarative API.
+
+### Resource usage
+The very long-term goal is clusterless Materialize, where Materialize does automatic workload scheduling for the customer.
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:50]: cc @frankmcsherry on this point in particular. We may want to try to clarify long-term (i.e, at least how many yaers away is it).

File: doc/developer/design/20231002_cluster_vision.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+An intermediary solution, which is also far off is autoscaling of clusters, where Materialize automatically resizes clusters based on the observed workload.
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:52]: I don't think this needs to be that far off! We could plausibly do this next year. Whereas I don't think clusterless Materialize is something we do in the next two years.

File: doc/developer/design/20231002_cluster_vision.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+A more achievable offering in the short-term is automatic shutdown of clusters, where Materialize can spin down a cluster to 0 replicas based on certain criteria, such as a scheduled time or amount of idle time. \
+This would reduce resource waste for development clusters. The triggering mechanism from graceful rehydration is also a requirement here.
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:55]: 👍🏽

File: doc/developer/design/20231002_cluster_vision.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+**Next**
+* Graceful rehydration, to support graceful manual execution of `ALTER...SET CLUSTER` and `ALTER...SET SIZE`.
+* Deprecate `CREATE/DROP CLUSTER REPLICA` for users.
+
+**Later**
+* Auto-shutdown of clusters.
+* Shadow replicas.
+
+**Much Later**
+* Autoscaling clusters / clusterless.
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:84]: I think we can promote autoscaling clusters to "later"!

File: doc/developer/design/20231002_cluster_vision.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+Support is able to create create unbilled or partially billed cluster resources for resolving customer issues. This is soon to be possible via unbilled replicas [#20317](https://github.com/MaterializeInc/materialize/issues/20317).
+
+### Roadmap
+**Now**
+* @antiguru to complete `ALTER...SET CLUSTER` [#20841](https://github.com/MaterializeInc/materialize/issues/20841), without graceful rehydration.
+* @antiguru to continue in-flight work on multipurpose clusters [#17413](https://github.com/MaterializeInc/materialize/issues/17413) - TODO(@antiguru): fill in details.
+* @ggnall to do discovery on the prescriptive data model as part of Blue/Green deployments project [#19748](https://github.com/MaterializeInc/materialize/issues/19748)
+
+**Next**
+* Graceful rehydration, to support graceful manual execution of `ALTER...SET CLUSTER` and `ALTER...SET SIZE`.
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:76]: Should this read "graceful reconfiguration"?

File: doc/developer/design/20231002_cluster_vision.md (hunk 1 above)
Comment (benesch) [reply doc/developer/design/20231002_cluster_vision.md]: No, I think it's good! TBH, I'd bring back the bit about shadow replicas and just add a caveat like "if they can be made safe." But I think it's absolutely right that we want _some_ way to test new releases/candidate changes on real production workloads, if we can find a way to do so without putting those environments at risk.

File: doc/developer/design/20231002_cluster_vision.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+### Declarative vs Imperative
+We should move toward a declarative API for managing clusters, where:
+
+Declarative is like `CREATE CLUSTER` with managed replicas and \
+Imperative is like `CREATE/DROP CLUSTER REPLICA`.
+
+This means deprecating manual cluster replica management. \
+We believe this is easier to use and manage.
+
+The primary work item for this is **graceful reconfiguration**. At the moment, a change in size causes downtime until the new replicas are hydrated. As such, customers still want the flexibility to create their own replicas for graceful resizing. We can avoid this by leaving a subset of the original replicas around until the new replicas are hydrated. \
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:42]: I'm increasingly wondering if we _don't_ need graceful reconfiguration. At least, if you buy into the theory that blue/green is you do things gracefully ... then perhaps `ALTER ... SET CLUSTER` _should_ be a dangerous forcing operation that puts things into action immediately, even if doing so causes downtime.

**Review by benesch — COMMENTED** [review]
This is really great, @chaas. Left a few comments within, but so excited that you've managed to capture all of this in writing.

File: doc/developer/design/20231002_cluster_vision.md (hunk 7 above)
Comment (chaas) [reply doc/developer/design/20231002_cluster_vision.md:42]: Ah kind of like what we're thinking for indexes, where if you want to in-place drop an index that will incur downtime, and if you want to gracefully drop one we suggest blue/green?

My only hesitation there is that forcing users to do blue/green is more work and conceptual overhead for the user than graceful reconfiguration. From a UX standpoint, graceful reconfiguration is effectively an abstraction on top blue/green, that manages spinning up the new resource and cutting over for them.
Even with blue/green, we'd need (1) to give users an easy way to detect that rehydration is complete so that they can execute the cutover themselves. We just wouldn't need (2), the triggering mechanism.

Also I was thinking the syntax here would be `ALTER CLUSTER ... SET (SIZE = <>)`, which spins up different sized replicas and cuts over to those replicas. As it stands now, adding another replica to an existing cluster doesn't require users to drop and recreate all downstream dependencies. My understanding is that blue/green requires recreating a whole second version of the stack, including all objects and clusters, which is a much heavier lift.

I think we should strive for graceful reconfiguration as an end state for resizing, with blue/green as an intermediary state.
For moving objects between clusters, I'm fine with that continuing to require blue/green.

Comment (benesch) [reply doc/developer/design/20231002_cluster_vision.md:42]: > Even with blue/green, we'd need (1) to give users an easy way to detect that rehydration is complete so that they can execute the cutover themselves.

Yep, definitely! But I think we're much closer to being able to surface that rehydration complete signal than we are to being able to take action off of it directly inside of Materialize. 

It's a good point that moving objects between clusters and scaling up a cluster are meaningfully different operations. Moving an object between clusters is likely part of your development workflow—e.g., maybe a big refactoring of how you map objects to clusters that you'll test extensively with a blue/green setup. But scaling up/down a cluster is likely something that you do live on the system (e.g., in response to load) and you both don't want that to cause downtime but it's overkill to reach for a `dbt`-managed blue/green deployment just to scale up/down.

File: doc/developer/design/20231002_cluster_vision.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+system.\
+In this workflow, downtime is not acceptable.\
+A command like `ALTER CLUSTER ... SET (SIZE = <>)` (resizing a cluster) would fall under this
+category.
+
+If a user wants to do a development workflow on a production system, they must use **blue/green
+deployments**. For example, if the user wants to move an object between clusters, they must use
+blue/green to set up another version of the object/cluster and cutover the production system
+to it once the object is rehydrated and ready.\
+Again, for this workflow, exposing hydration status is the primary work item.
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:70]: This is a great framing. Thank you!

File: doc/developer/design/20231002_cluster_vision.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+<!--
+What other solutions were considered, and why weren't they chosen?
+
+This is your chance to demonstrate that you've fully discovered the problem.
+Alternative solutions can come from many places, like: you or your Materialize
+team members, our customers, our prospects, academic research, prior art, or
+competitive research. One of our company values is to "do the reading" and
+to "write things down." This is your opportunity to demonstrate both!
+-->
```
Comment (benesch) [inline doc/developer/design/20231002_cluster_vision.md:160]:
```suggestion
```

I'd just kill these sections? I don't think they make a ton of sense for a vision-style design document.

**Review by benesch — COMMENTED** [review]
I think this looks great, @chaas! Anything on your end that you're keeping it in draft for? When you're back, I think this is ready for wider circulation!

**Comment by chaas** [comment]
> Here are two examples that require replicas with different sizes: For instance, it's an easy way to verify the effect of scaling up/down clusters. Right now, a customer I'm working with is seeing hydration times around 30min but the replica can potentially be scaled down. Adding an additional smaller replica is a very seamless way to just test what happens without actually scaling the cluster. Similarly, cost cautious customers may be willing to run replicas pretty busy, if they want to verify upfront if a smaller replica does not fall over, they can provision the smaller replica in addition to the larger one and monitor for some time (hours or few days) what happens.
> 
> People may be able to do similar things with automatic and 0 down time scaling. I'm not saying that we should not focus on the declarative approach. But keeping some basic support for the more manual imperative approach could be useful for some specific and less common use cases.

@sthm Thanks for sharing!
The way to do this with the declarative API would be blue/green deployments, where they set up a parallel version of the objects & cluster in whatever size they want to experiment with as much as they want, without risking impacting their production workflow. However, I see how that's gonna be heavyweight and complicated for users trying to do simple things.
I will brainstorm some alternative ideas here...if it's a common enough use case, perhaps we can add syntactic sugar that allows for very specific things like "test with another size", that abstracts away the underlying replica architecture.
Allowing users to directly manipulate the replicas to run that kind of experimentation on their production environment does feel a little "scary", since it e.g. increases the risk of user error like deleting the wrong replica.

File: doc/developer/design/20231002_cluster_vision.md (hunk 9 above)
Comment (chaas) [reply doc/developer/design/20231002_cluster_vision.md:160]: agreed, sounds good

**Comment by benesch** [comment]
> perhaps we can add syntactic sugar that allows for very specific things like "test with another size", that abstracts away the underlying replica architecture.

Just for posterity: the user-facing "replica sets" that Moritz proposed is another way out, in that we could tell users who want to test with multiple replica sizes on the same cluster to create multiple replica sets.

File: doc/developer/design/20231002_cluster_vision.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+### Support & testing
+Support is able to create create unbilled or partially billed cluster resources for resolving customer issues. This is soon to be possible via unbilled replicas [#20317](https://github.com/MaterializeInc/materialize/issues/20317).
+
+Engineering may also want the ability to create unbilled shadow replicas for testing new features and
+query plan changes, which do not serve customers' production workflows, if they can be made safe.
+
+### Roadmap
+**Now**
+* @antiguru to complete `ALTER...SET CLUSTER` [#20841](https://github.com/MaterializeInc/materialize/issues/20841), without graceful rehydration.
```
Comment (antiguru) [inline doc/developer/design/20231002_cluster_vision.md:108]: Reference #17417 instead. The effort is currently blocked on figuring out the right design.

File: doc/developer/design/20231002_cluster_vision.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+* @antiguru to continue in-flight work on multipurpose clusters [#17413](https://github.com/MaterializeInc/materialize/issues/17413) - TODO(@antiguru): fill in details.
```
Comment (antiguru) [inline doc/developer/design/20231002_cluster_vision.md:109]: I think my remaining work item was #21846, which enables co locating compute and storage objects. It's behind a feature flag, pending changes to storage to turning it on for select users.

File: doc/developer/design/20231002_cluster_vision.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+* Enable fault tolerance/use-case isolation
+
+### Declarative vs Imperative
+We should move toward a declarative API for managing clusters, where:
+
+Declarative is like `CREATE CLUSTER` with managed replicas and \
+Imperative is like `CREATE/DROP CLUSTER REPLICA`.
+
+This means deprecating manual cluster replica management. \
+We believe this is easier to use and manage.
```
Comment (frankmcsherry) [inline doc/developer/design/20231002_cluster_vision.md:42]: Strong disagree here. There's maybe a false dichotomy at play, as there is a middle ground between "deprecate manual cluster management" and "default to manual cluster management". As long as MZ has downtime on a thing that could have been done manually, it's a real hard sell that we should forbid doing the manual thing (e.g. resizing).

An alternative would be "teach people to type `ALTER CLUSTER REPLICAS` rather than `CREATE CLUSTER REPLICA` and `DROP CLUSTER REPLICA`", which is six of one half dozen of another to me. Still mostly imperative (a human types a command, just about the goal state rather than the transition) but with less cognitive overhead. But stops short of "no manual replica management".

Comment (frankmcsherry) [reply doc/developer/design/20231002_cluster_vision.md:42]: If nothing else, it would be helpful to unpack the intended "imperative" vs "declarative" distinction. SQL's command language, for example, is painfully *imperative* and not at all declarative. But it's hard for me to understand at this point what the distinction is other than removing a user's ability to control the assignment of their money (in the form of replicas) to their work.

File: doc/developer/design/20231002_cluster_vision.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+In a production workflow, the underlying set of objects are actively depended on by a production
+system.\
+In this workflow, downtime is not acceptable.\
+A command like `ALTER CLUSTER ... SET (SIZE = <>)` (resizing a cluster) would fall under this
+category.
+
+If a user wants to do a development workflow on a production system, they must use **blue/green
+deployments**. For example, if the user wants to move an object between clusters, they must use
+blue/green to set up another version of the object/cluster and cutover the production system
+to it once the object is rehydrated and ready.\
```
[the remark this answers is not in the document]
Comment (frankmcsherry) [reply doc/developer/design/20231002_cluster_vision.md:69]: Causing work to happen on a cluster, for example in response to `CREATE INDEX`, isn't fundamentally different from what happens when you type `SELECT .. FROM not_an_index`; I'm unclear if the blue/green discipline is actually a "no creating dataflows" discipline, one outcome of which is forced blue/green modification but it derives from something more specific.

File: doc/developer/design/20231002_cluster_vision.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+Again, for this workflow, exposing hydration status is the primary work item.
+
+For production workflows, like resizing an active cluster, blue/green is an acceptable intermediate
+solution, but is an overkill amount of work for such a simple action.
```
Comment (frankmcsherry) [inline doc/developer/design/20231002_cluster_vision.md:73]: This text is confusing to me! What would it mean to blue/green a resized cluster? Like, the act of resizing would amount to creating a new cluster, with different resources behind it, and cutting over from one to the other? It is hard for me to understand this in the context of a blue/green implementation that e.g. rebuild and renames things, where downstream dependents are left confused. Would we drop sinks when we do this, for example?

**Review by frankmcsherry — COMMENTED** [review]
I left a bunch of comments, but I didn't get a lot of clarity out of the document about the long term vision for clusters. I think part of this is that there is some background context you have that isn't written down here (what specifically you mean by blue/green, for example, which is meant to solve several problems but either does/doesn't depending on exactly what is meant by it), as well as some unexplored content (to what extent does affecting cluster definitions have deleterious consequences for downstream users).

I wrote some text that I thought was clarifying for me, and I'm curious if it is also clarifying for other readers of the doc. It doesn't have a prescriptive take, but I think lays out the options in a way that are clearer for me. Specifically, it distinguishes between cluster *responsibilities* and cluster *provisioning*, and how each of them might evolve (potentially independently). 

---


Clusters are definitions of bundles of related but otherwise isolated work a user has framed.
Clusters are initially empty, but evolve both in their *responsibilities* (work they must perform) and their *provisioning* (which resources have been allocated to the work). 
The also evolve with time, in that cluster changes may put them in transient states where their provisioned resources are not all up to date with the work they must perform, and with time (ideally) they manage to catch up.

The commands on clusters are split between modifying cluster responsibilities, and modifying cluster provisioning.

There are two ways that we will recommend folks modify cluster responsibilities: *in situ* or *blue/green*. 
1.  In situ modifications change an existing cluster definition, which can be prompt but is also very risky.
    If adding a resource brings the cluster over some limit, the cluster may collapse.
    Without bringing the cluster up from zero, we are less certain that the modified cluster responsibilities can be brought up at all.
    We believe this is a fine mode for development operation, where fast iteration is important and the user can tolerate downtime (e.g. hydration, failure).
2.  Blue/green modifications clone a cluster definition and modify the clone, which takes time to have effect but can expose problems before deploying them.
    Blue/green could be enforced by putting it in a state that disallows cluster modification.
    Ergonomically, it feels like we would want a `COPY CLUSTER` command, or something analogous that prepares a mutable cluster definition without forcing an undeploy of the existing cluster.
    We would also potentially want a `SWAP CLUSTER` command, which would have the effect of installing the new definition in place of the old definition.
    Such a `SWAP` command could incur downtime if the new cluster is not yet in a comparably healthy state. 

There are multiple ways that we will recommend folks modify cluster *provisioning*: 
1.  Imperative (manual) modification of replicas backing a cluster. 
    This is a robust fall-back in cases where we have yet to invent a policy that implements user goals.
2.  Declarative definitions of replicas backing a cluster.
    Examples here include
        a. Automated shut-down during inactivity, or outside of business hours.
        b. Automated scaling if resource use is near maximumum or minimum thresholds. 
        c. Automated replication to multiple availability zones.
    The list seems plausibly long and nuanced, and probably indefinitely incomplete. 
3.  Immutable cluster provisioning.
    Cluster provisioning is locked at cluster creation, and to change it requires a blue/green sort of redeployment.

File: doc/developer/design/20231002_cluster_vision.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ggnall) [reply doc/developer/design/20231002_cluster_vision.md:69]: I think your text above is accurate here! The idea is that we don't advise users to modify cluster responsibilities on a cluster that's receiving queries from a production consumer. 
>But also, I'm maybe confused about what "doing a development workflow on a production system" means.

One might YOLO CREATE INDEX an an index to a production cluster, particularly to fight an emergency, but we might not want to build shortcuts for them to move indexes around production clusters outside a that flow.

Related, I would argue that we need to be more vs. less prescriptive on this workflows! Regardless of whether this is the perfect take or not.
>This sounds like a very prescriptive take that I could imagine folks chafing against.

File: doc/developer/design/20231002_cluster_vision.md (hunk 14 above)
Comment (chaas) [reply doc/developer/design/20231002_cluster_vision.md:73]: Agreed - Frank and I discussed offline and blue/green is too burdensome for scaling use-cases, and particularly with the versions of blue/green that we will realistically be building now which will be manually-controlled.

File: doc/developer/design/20231002_cluster_vision.md (hunk 2 above)
Comment (antiguru) [reply doc/developer/design/20231002_cluster_vision.md:84]: I'm fine with it staying as it currently is, i.e., we use clusters as a user-indicated boundary between resources. One problem that I'd eventually like to see vanish is how do users determine the right cut in their dependency graph such that they can use the least amount of resources while achieving their availability goals. From what I observed, this is a recurring problem which needs some explaining for users to get right.

How we get there is a different question. One take could be that there's something that indicates a resource assignment, but I have no strong preference whether this would be part of a component within Materialize or something on top only giving recommendations. The latter seems more practical and potentially less dangerous, at least until we figure out how to write a controller for Materialize (which we currently don't know.)

TL;DR, happy to delay this infinitely, but we should be aware of the challenge users face.
